### PR TITLE
Added secure strlen with memchr.

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/string.h [hunk 1]
```diff
@@ -205,6 +205,12 @@ struct aws_byte_cursor aws_byte_cursor_from_string(const struct aws_string *src)
 AWS_COMMON_API
 struct aws_string *aws_string_clone_or_reuse(struct aws_allocator *allocator, const struct aws_string *str);
 
+/* Computes the length of a c string in bytes assuming the character set is either ASCII or UTF-8. If no NULL character
+ * is found within max_read_len of str, AWS_ERROR_C_STRING_BUFFER_NOT_NULL_TERMINATED is raised. Otherwise, str_len
+ * will contain the string length minus the NULL character, and AWS_OP_SUCCESS will be returned. */
+AWS_COMMON_API
+int aws_c_string_secure_strlen(const char *str, size_t max_read_len, size_t *str_len);
```
Comment (ColdenCullen) [inline include/aws/common/string.h]: Just to bikeshed on naming a bit, I think having "c_string" and "strlen" in the name is redundant. Maybe just `aws_secure_strlen` (matches `aws_secure_zero`) or `aws_c_string_secure_length` (matches stuff in this file) instead?

Comment (JonathanHenson) [reply include/aws/common/string.h]: sure

Comment (graebm) [reply include/aws/common/string.h]: +1 aws_secure_strlen()
or just aws_strlen()

File: source/string.c [hunk 2]
```diff
@@ -258,3 +258,21 @@ struct aws_string *aws_string_clone_or_reuse(struct aws_allocator *allocator, co
     AWS_POSTCONDITION(aws_string_is_valid(str));
     return aws_string_new_from_string(allocator, str);
 }
+
+int aws_c_string_secure_string_length(const char *str, size_t max_read_len, size_t *str_len) {
+    if (!str || !str_len) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
```
Comment (feliperodri) [inline source/string.c]:
```
AWS_ERROR_PRECONDITION(!str || !str_len, AWS_ERROR_INVALID_ARGUMENT)
```

File: include/aws/common/string.h [hunk 3]
```diff
@@ -205,6 +205,12 @@ struct aws_byte_cursor aws_byte_cursor_from_string(const struct aws_string *src)
 AWS_COMMON_API
 struct aws_string *aws_string_clone_or_reuse(struct aws_allocator *allocator, const struct aws_string *str);
 
+/* Computes the length of a c string in bytes assuming the character set is either ASCII or UTF-8. If no NULL character
+ * is found within max_read_len of str, AWS_ERROR_C_STRING_BUFFER_NOT_NULL_TERMINATED is raised. Otherwise, str_len
+ * will contain the string length minus the NULL character, and AWS_OP_SUCCESS will be returned. */
+AWS_COMMON_API
+int aws_c_string_secure_string_length(const char *str, size_t max_read_len, size_t *str_len);
```
Comment (justinboswell) [inline include/aws/common/string.h]: I thought this was going to be renamed ```aws_secure_strlen()```?

Comment (JonathanHenson) [reply include/aws/common/string.h]: I'll rename it.

**Pull request merged**